jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -664,213 +175,117 @@ def mk_freeze_script(pub, locktime):
-    this tx input.
-    Note on 'script' and 'scriptCode':
-    For p2pkh, 'script' should be the output script (scriptPubKey), and
-    the scriptCode should be left as the default None.
-    For p2sh non-segwit multisig, the script should be the
-    output of mk_multisig_script (and scriptCode and amount None).
-    For either nested (p2sh) or not p2wpkh and p2wsh, the scriptCode is the
-    preimage of the scriptPubKey hash (the witness program, which here
-    has been passed in in the 'script' parameter).
-
-    Note that for the p2sh-p2wsh and p2wsh cases, a redeem script that is not
-    multisig should still be correctly handled here; the codebase restricts
-    the creation of such signatures to multisig only, but *verification* should
-    function correctly with any custom script.
+    return CScript([btc.OP_RETURN, data])
 
-    TODO: Also note that the OP_CODESEPARATOR special case outlined in BIP143
-    is not yet covered.
+def sign(tx, i, priv, hashcode=SIGHASH_ALL, amount=None, native=False):
     """
-    assert isinstance(i, int)
-    assert i >= 0
-    if not isinstance(tx, bytes):
-        tx = binascii.unhexlify(tx)
-    if not isinstance(script, bytes):
-        script = binascii.unhexlify(script)
-    if scriptCode is not None and not isinstance(scriptCode, bytes):
-        scriptCode = binascii.unhexlify(scriptCode)
-    if isinstance(sig, bytes):
-        sig = binascii.hexlify(sig).decode('ascii')
-    if isinstance(pub, bytes):
-        pub = binascii.hexlify(pub).decode('ascii')
-
-    hashcode = binascii.unhexlify(sig[-2:])
-
-    if amount:
-        modtx = segwit_signature_form(deserialize(tx), i,
-                    scriptCode, amount, hashcode, decoder_func=lambda x: x)
-    else:
-        modtx = signature_form(tx, i, script, hashcode)
-    return ecdsa_tx_verify(modtx, sig, pub, hashcode)
-
-
-def sign(tx, i, priv, hashcode=SIGHASH_ALL, usenonce=None, amount=None,
-         native=False):
-    """
-    Given a serialized transaction tx, an input index i, and a privkey
-    in bytes or hex, returns a serialized transaction, in hex always,
-    into which the signature and/or witness has been inserted. The field
-    `amount` flags whether segwit signing is to be done, and the field
-    `native` flags that native segwit p2wpkh signing is to be done. Note
-    that signing multisig is to be done with the alternative functions
-    get_p2sh_signature or get_p2wsh_signature (and non N of N multisig
-    scripthash signing is not currently supported).
+    Given a transaction tx of type CMutableTransaction, an input index i,
+    and a raw privkey in bytes, updates the CMutableTransaction to contain
+    the newly appended signature.
+    Only four scriptPubKey types supported: p2pkh, p2wpkh, p2sh-p2wpkh, p2wsh.
+    Note that signing multisig must be done outside this function, using
+    the wrapped library.
+    If native is not the default (False), and if native != "p2wpkh",
```

Comment:
This is not good style, because `native` could be three different types, which really harms how easy it is to read and understand the code. How about renaming the variable to something like `input_script_type`, and perhaps removing the option to be `False` and instead make that the string `"p2sh-p2wpkh"`.

Reply:
Speaking of style, I would say that using `Enum` instead of strings to denote scriptPubKey types would be better (and could help catch bugs, especially if static checker like mypy is used). But at the same time I think that this PR is already huge enough to consider deferring any non-essential changes to the later PRs

Reply:
I'd argue that the long-term success of a project rests (among other things) on its maintainability, and therefore stuff like this is essential.

Let's be honest, if its deferred now it will probably never happen. And I think the change is quite short, it only needs to change the default value `native="p2sh-p2wpkh"` and then in the function body change `if native:` to `if native == "p2sh-p2wpkh":`. Renaming `native` to `input_script_type` could be deferred I guess, but its doable with a find-and-replace. Using an `Enum` is definitely a good idea as well but might take a bit longer to do (but maybe not, it also just needs a find-and-replace). The automated tests are there to help us make sure no bugs appear because of this.

Reply:
By 'essential' here I meant essential for the goal of this particular PR. In the whole-project context, of course such things would be essential. To prevent "if its deferred now it will probably never happen" in cases like this, I usually create an issue in the issue tracker to ensure the problem is visible, and the change will eventually happen. Of course it is up to the author of this PR to decide the scope of the changes that is suitable and if the size of the PR is still manageable, and my comment about deferring the changes was probably uncalled for, sorry for this extra noise.

Reply:
> This is not good style, because `native` could be three different types, which really harms how easy it is to read and understand the code. How about renaming the variable to something like `input_script_type`, and perhaps removing the option to be `False` and instead make that the string `"p2sh-p2wpkh"`.

This is certainly a good point. Note that such a janky architectural decision doesn't come entirely out of ignorance, it comes mostly out of pragmatism: I am *very* careful/worried/conservative when it comes to changing this part of the code, so when I realised that (a) in principle it is *very* simple to encompass additional transaction types (here, p2wsh), that is to say, almost no additional code in the body of the function (just replacing the scriptCode field in the sighash operation), but that (b) we have to change the function signature, i just tried to make the change as minimal as possible. That (b) is, I agree, not at all ideal.

So overall I definitely agree with your intention: make the function signature more constrained (duck typing is not great here!), and make `native=` have only one type. But the suggestions in the above don't quite work. They can't all be a string, the p2wsh case must be a CScript (or bytes) instance. I think the most strict/correct way to do it is:

```python3 
def sign(tx, i, priv, hashcode=SIGHASH_ALL, amount=None, scriptpubkey_type="p2sh-p2wpkh", redeem_script=None):
```

This way the code will pull in the redeemscript for the p2wsh case if `scriptpubkey_type == "p2wsh"` (as it does now; it could also work for p2sh (general/multisig) or for p2sh-p2wsh but we explicitly have no support for those at the moment).

Also as per comments above scriptpubkey_type doesn't have to be "str" but just trying to keep it simple for now.